### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by dcermak**
Resurrection of #18, but includes the changes from #22 as well, as the CI will otherwise fail.

### CODE REVIEW

**Review by okurz — APPROVED** [review]

File: tests/test_client.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def test_add_auth_headers(self, simple_config):
         authed = client._add_auth_headers(prepared)
         assert prepared.headers != authed.headers
         assert authed.headers["X-API-Hash"] == "71373f0a57118b120d1915ccc0a24ae2cc112ad3"
-        assert authed.headers["X-API-Microtime"] == b"1582761600.0"
+        assert authed.headers["X-API-Microtime"] == "1582761600.0"
```
Comment (AdamWill) [inline tests/test_client.py:100]: is this...okay? Are we sure this works? I'm pretty sure I wrote this to behave the same as the perl client, so this just worries me a bit.

**Comment by AdamWill** [comment]
I merged #22 , so can you rebase this? Thanks.

I really hate the `@overload` stuff, just visually, but I don't see a way to lose it while maintaining compatibility sadly...

File: tests/test_client.py (hunk 1 above)
Comment (dcermak) [reply tests/test_client.py:100]: I am honestly not sure, @okurz do you have an idea? The issue is that requests' typing complains that `bytes` are not supposed to be in the header. But that could be wrong of course.

**Comment by dcermak** [comment]
> I merged #22 , so can you rebase this? Thanks.

Done

> I really hate the `@overload` stuff, just visually, but I don't see a way to lose it while maintaining compatibility sadly...

The `@overload` stuff is not strictly necessary, the type hints work without it. It just provides the additional benefit that the type checker can infer more precise return types if it knows some parameters "at compile time" (i.e. when scanning your code). E.g. if you write `do_request(…, parse=False)`, mypy will now know that the return type is a `requests.Response`. If we omit the overloads, then it would only know that it's either a `requests.Response` or `Any`. It's not overly useful in this case, so if it bothers you, we can remove it entirely.

File: tests/test_client.py (hunk 1 above)
Comment (okurz) [reply tests/test_client.py:100]: I don't know. But can you exclude this from the typing check somehow?

**Comment by AdamWill** [comment]
I merged some other PRs, so can you rebase this again? Sorry. I want to check it out and see if authed requests actually work with it before merging, also check the tests still pass. Note, I've renamed the default branch to 'main' so you'll have to change that on your local checkout.

**Comment by okurz** [comment]
@Mergifyio rebase

**Comment by dcermak** [comment]
@AdamWill done. I also had to revert https://github.com/dcermak/openQA-python-client/commit/f2db9d08dacefd3581f5c546e5c4af1f158d9fbd as that set the comparison branch to `origin/master` which is now gone.

**Comment by AdamWill** [comment]
@dcermak  It just needs to be changed to origin/main. Sorry, thought I did that, I'll do it on main branch now and we can rebase this again...

**Comment by AdamWill** [comment]
Done, can you rebase again and leave `tox.ini` alone, hopefully it should work? Thanks.

**Comment by dcermak** [comment]
Adam Williamson ***@***.***> writes:

> Done, can you rebase again and leave `tox.ini` alone, hopefully it should work? Thanks.

I have dropped by revert and rebased on main.

**Pull request merged**